compiler/generator_go.c

```diff
@@ -0,0 +1,1253 @@
+
+    struct str * loopvar = vars_newname(g);
+    write_comment(g, p);
+    w(g, "~Mfor _ = range make([]struct{},");
+    generate_AE(g, p->AE);
+    writef(g, ") {~+~N", p);
+
+    generate(g, p->left);
+
+    w(g, "~-~M}~N");
+    str_delete(loopvar);
+    g->unreachable = false;
+}
+
+static void generate_repeat(struct generator * g, struct node * p, struct str * loopvar) {
+
+    struct str * savevar = vars_newname(g);
+    int keep_c = repeat_restore(g, p->left);
+    int replab = new_label(g);
+    g->I[0] = replab;
+    write_comment(g, p);
+    writef(g, "~Mreplab~I0: for{~N~+", p);
+
+    if (keep_c) write_savecursor(g, p, savevar);
+
+    g->failure_label = new_label(g);
+    str_clear(g->failure_str);
+    g->I[0] = g->failure_label;
+    w(g, "~Mlab~I0: for _ = range [2]struct{}{} {~N~+");
+    generate(g, p->left);
+
+    if (!g->unreachable) {
+        if (loopvar != 0) {
+            g->B[0] = str_data(loopvar);
+            w(g, "~M~B0--~N");
+        }
+
+        g->I[0] = replab;
+        w(g, "~Mcontinue replab~I0~N");
+    }
+    w(g, "~-~M}~N");
+    g->unreachable = false;
+
+    if (keep_c) write_restorecursor(g, p, savevar);
+
+    g->I[0] = replab;
+    w(g, "~Mbreak replab~I0~N~-~M}~N");
+    str_delete(savevar);
+}
+
+static void generate_atleast(struct generator * g, struct node * p) {
+
+    struct str * loopvar = vars_newname(g);
+    write_comment(g, p);
+    g->B[0] = str_data(loopvar);
+    w(g, "~Mvar ~B0 = ");
+    generate_AE(g, p->AE);
+    w(g, "~N");
+    {
+        int a0 = g->failure_label;
+        struct str * a1 = str_copy(g->failure_str);
+
+        generate_repeat(g, p, loopvar);
+
+        g->failure_label = a0;
+        str_delete(g->failure_str);
+        g->failure_str = a1;
+    }
+    g->B[0] = str_data(loopvar);
+    write_failure_if(g, "~B0 > 0", p);
+    str_delete(loopvar);
+}
+
+static void generate_setmark(struct generator * g, struct node * p) {
+
+    write_comment(g, p);
+    g->V[0] = p->name;
+    writef(g, "~M~V0 = env.Cursor~N", p);
+}
+
+static void generate_tomark(struct generator * g, struct node * p) {
+
+    write_comment(g, p);
+    g->S[0] = p->mode == m_forward ? ">" : "<";
+
+    w(g, "~Mif env.Cursor ~S0 "); generate_AE(g, p->AE);
+    writef(g, " ", p);
+    write_block_start(g);
+    write_failure(g);
+    write_block_end(g);
+    g->unreachable = false;
+    w(g, "~Menv.Cursor = "); generate_AE(g, p->AE); writef(g, "~N", p);
+}
+
+static void generate_atmark(struct generator * g, struct node * p) {
+
+    write_comment(g, p);
+    w(g, "~Mif env.Cursor != "); generate_AE(g, p->AE);
+    writef(g, " ", p);
+    write_block_start(g);
+    write_failure(g);
+    write_block_end(g);
+    g->unreachable = false;
+}
+
+static void generate_hop(struct generator * g, struct node * p) {
+
+    write_comment(g, p);
+    g->S[0] = p->mode == m_forward ? "" : "-";
+
+    w(g, "~Mvar c = env.ByteIndexForHop(~S0");
```

Comment:
It would be great to have one, but currently I'm lacking the time to sort it out myself.

`hop -1` is equivalent to `false` so rather an artificial example, but `hop -$x` is something you might more realistically encounter.